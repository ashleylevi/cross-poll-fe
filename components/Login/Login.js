import React from 'react';
import { StyleSheet, Text, View, TextInput, Button, TouchableOpacity } from 'react-native';
import { RadioButtons } from 'react-native-radio-buttons'
import { Icon } from 'react-native-elements';

export default class Login extends React.Component {
  constructor(props) {
    super(props)
    this.state = {
      newUser: true,
      email: '',
      firstName: '',
      lastName: '',
      password: '',
      passwordConfirmation: '',
      error: true,
      errorMessage: '',
      selectedOption: 'sign up'
    }
  }

  submitUser = () => {
    if(this.state.selectedOption === 'sign up') {
      this.submitNewUser()
    } else {
      this.submitExistingUser()
    }
  }

  submitExistingUser = () => {
    const { email, password } = this.state;
    const postBody = {
      email,
      password
    }
    fetch('https://adoptr-be.herokuapp.com/api/v1/sessions', {
      method: 'POST',
      body: JSON.stringify(postBody),
      headers:{
        'Content-Type': 'application/json'
      }
    })
    .then( response => response.json())
    .then( result => this.checkForError(result))
    .catch(error => console.log(error))
  }

  checkForError = (result) => {
    if(result.error) {
      this.setState({
        error: true,
        errorMessage: result.error
      })
    } else {
      this.props.updateUserToken(result.data.attributes.api_token)
    }
  }

  submitNewUser = () => {
    const { firstName, lastName, email, password, passwordConfirmation } = this.state;
    const postBody = {
      first_name: firstName,
      last_name: lastName,
      email,
      password,
      password_confirmation: passwordConfirmation
    }
    console.log(postBody)
    fetch('https://adoptr-be.herokuapp.com/api/v1/users', {
      method: 'POST',
      body: JSON.stringify(postBody),
      headers:{
        'Content-Type': 'application/json'
      }
    })
    .then( response => response.json())
    .then( result => this.checkForError(result))
    .catch(error => console.log(error))
  }
  
  toggleLogin = () => {
    this.setState({
      newUser: false,
      error: false,
      errorMessage: '',
      email: '',
      firstName: '',
      lastName: '',
      password: '',
      passwordConfirmation: ''
    })
  }

  toggleSignUp = () => {
    this.setState({
      newUser: true,
      error: false,
      errorMessage: '',
      email: '',
      firstName: '',
      lastName: '',
      password: '',
      passwordConfirmation: ''
    })
  }

  setSelected = (selectedOption) => {
    this.setState({
      selectedOption
    })
    this.setState({
      error: false,
      errorMessage: '',
      email: '',
      firstName: '',
      lastName: '',
      password: '',
      passwordConfirmation: ''
    })
  }

  renderOption = (option, selected, onSelect, index) => {
    const selectedRadio = selected ? styles.selectedRadio : styles.unselectedRadio
    const selectedText = selected ? styles.selectedText : styles.unselectedText
    return (
      <TouchableOpacity onPress={onSelect} key={index} style={selectedRadio} >
        <Text style={selectedText}>{option}</Text>
      </TouchableOpacity>
    )
  }

  renderContainer = (optionNodes) => {
    return (<View style={styles.radioButtons}>
        {optionNodes}
          </View>)
  }

  render() {
    const selectOptions = ["sign in", "sign up"]

    return (
      <View style={styles.form}>
        <View style={styles.title}> 
          <Text style={styles.titleText}> 
            AdoptR 
          </Text>
          <Icon
              name='paw'
              type='font-awesome'
              color='white'
              size= {48}
              iconStyle={styles.pawprint}
            />
        </View>
        <View style={this.state.selectedOption === 'sign up' ? styles.formSignUp : styles.formSignIn}>
          <RadioButtons
                style={styles.radioButtons}
                options={ selectOptions }
                onSelection={ this.setSelected }
                selectedOption={this.state.selectedOption }
                renderOption={ this.renderOption }
                renderContainer={ this.renderContainer }
              />
        <TextInput 
          style={this.state.selectedOption === 'sign up' ? styles.input : styles.hidden}
          placeholder='first name'
          autoCapitalize={"none"}
          value={this.state.firstName}
          onChangeText={(value) => this.setState({firstName: value})}
        />
        <TextInput 
          style={this.state.selectedOption === 'sign up' ? styles.input : styles.hidden}
          placeholder='last name'
          autoCapitalize={"none"}
          value={this.state.lastName}
          onChangeText={(value) => this.setState({lastName: value})}
        />
        <TextInput 
          type='email'
          autoCapitalize={"none"}
          style={styles.input}
          placeholder='email'
          value={this.state.email}
          onChangeText={(value) => this.setState({email: value})}
        />
        <TextInput
          style={styles.input}
          secureTextEntry={true}
          placeholder='password'
          autoCapitalize={"none"}
          value={this.state.password}
          onChangeText={(value) => this.setState({password: value})}
        />
        <TextInput
          style={this.state.selectedOption === 'sign up' ? styles.input : styles.hidden}
          secureTextEntry={true}
          placeholder='confirm password'
          value={this.state.passwordConfirmation}
          onChangeText={(value) => this.setState({passwordConfirmation: value})}
        />
        <Text style={this.state.error ? styles.errorMessage : styles.hidden}>
          {this.state.errorMessage}
        </Text>
        </View>
        <TouchableOpacity
        style={styles.button}
        onPress={()=>this.submitUser()} > 
          <Text style={styles.submitButtonText}  > 
          submit 
          </Text>
        </TouchableOpacity>
      </View>
    )
  }
}

const styles = StyleSheet.create({
  buttonText: {
    fontSize: 20
  },
  form: {
    backgroundColor: '#E5E5E5',
    fontSize: 50,
    width: '100%',
    flex: 1,
    alignItems: 'center',
    justifyContent: 'space-evenly',
    paddingTop: 160,
    paddingBottom: 140
  },
  formSignUp: {
    backgroundColor: 'white',
    fontSize: 60,
    width: '96%',
    flex: 1,
    alignItems: 'center',
    justifyContent: 'space-around',
    borderRadius: 16,
    margin: 10,
    paddingTop: 60,
    paddingBottom: 100,
    shadowColor: 'black',
    shadowOffset: { width: 1, height: 2 },
    shadowOpacity: .8,
    shadowRadius: 4,
    // paddingBottom: 100
  },
  formSignIn: {
    // justifyContent: 'space-around',
    // paddingBottom: 100
    alignItems: 'center',
    backgroundColor: 'white',
    borderRadius: 16,
    flex: 1,
    fontSize: 60,
    height: 200,
    margin: 10,
    paddingBottom: 100,
    paddingTop: 100,
    width: '96%',
    shadowColor: 'black',
    shadowOffset: { width: 1, height: 2 },
    shadowOpacity: .8,
    shadowRadius: 4,
  },
  input: {
    // backgroundColor: 'rgba(255,255,255, 0.8)',
    borderBottomColor: '#048BA8',
    // borderRadius: 24,
    borderStyle: 'solid',
    borderBottomWidth: 2,
    fontSize: 20,
    height: 66,
    // shadowColor: '#000',
    // shadowOffset: { width: 1, height: 2 },
    // shadowOpacity: .8,
    // shadowRadius: 4,
    textAlign: 'left',
    width: 320,
    fontFamily: 'Kohinoor Bangla'
  },
  title: {
    borderRadius: 24,
    shadowOffset: { width: 0, height: 1 },
    shadowOpacity: 0.8,
    shadowRadius: 2,
    alignItems: 'center',
    backgroundColor: '#048BA8',
    color: 'white',
    flex: 1,
    flexDirection: 'row',
    fontFamily: 'Kohinoor Bangla',
    fontSize: 50,
    justifyContent: 'center',
    position: 'absolute',
    padding: 10,
    shadowColor: 'black',
    textAlign: 'center',
    top: 60,
    zIndex: 3,
  },
  titleText: {
    fontFamily: 'Kohinoor Bangla',
    fontSize: 52,
    textAlign: 'center',
    color: 'white',
    // margin: 4,
  },
  pawprint: {
    color: 'white',
    marginLeft: 20
  },
  hidden: {
    display: 'none'
  },
  button: {
    backgroundColor: '#048BA8',
    borderColor: 'black',
    borderStyle: 'solid',
    borderRadius: 24,
    shadowColor: 'black',
    shadowOffset: { width: 1, height: 2 },
    shadowOpacity: .8,
    shadowRadius: 4,
    // borderWidth: 2,
    width: 300,
    position: 'absolute',
    bottom: 50,
  },
  submitButtonText: {
    fontSize: 32,
    color: 'white',
    textAlign: 'center',
    fontFamily: 'Kohinoor Bangla'
  },
  selectedText: {
    fontWeight: 'bold',
    color: '#048BA8',
    textDecorationLine: 'underline',
    textDecorationColor: '#048BA8',
    fontSize: 32,
    fontFamily: 'Kohinoor Bangla'
  },
  unselectedText: {
    fontSize: 32,
    fontFamily: 'Kohinoor Bangla'
  },
  radioButtons: {
    display: 'flex',
    flexDirection: 'row',
    width: 260,
    justifyContent: 'space-between',
    alignItems: 'center',
    position: 'absolute',
    top: 10,
    zIndex: 2
  },
  errorMessage: {
    color: 'red',
    fontSize: 20,
    textAlign: 'center',
    position: 'absolute',
    bottom: 16
  }
});