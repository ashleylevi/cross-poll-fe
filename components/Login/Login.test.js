import React from 'react'; 
import { shallow, mount } from 'enzyme';
import Login from '../Login/Login';
import renderer from 'react-test-renderer';

describe('Login', () => {
  let wrapper;

  beforeEach(() => {
    wrapper = shallow( <Login
      showLogin={true}
      updateUserToken={jest.fn()}
    />);
  })

  it('should match the snapshot with all data passed in correctly', () => { 
    expect(wrapper).toMatchSnapshot(); 
  })

  it.skip('should call the submit user func when clicked', () => { 
    const submitUserMock = jest.fn();
  
    let submitUser = wrapper.instance().submitUser()
    wrapper.find('Button').simulate('Press');
    expect(submitUser).toBeCalled(); 
  }); 

  it('should update state and reset error message when setSelected is called', () => {
    wrapper.instance().setSelected('sign up');
    expect(wrapper.state('selectedOption')).toEqual('sign up');
    expect(wrapper.state('error')).toEqual(false);
    expect(wrapper.state('errorMessage')).toEqual('');
  })

  it('should update state when name/email is typed in', () => {
  
  })

  describe('submitExistingUser', () => {
    it.skip('calls fetch with the correct data when logging in', () => { 

      const mockUser = {
        email: 'ashplevi@yahoo.com',
        password: 'password'
      }
      window.fetch = jest.fn().mockImplementation(() => Promise.resolve({
        json: () => Promise.resolve({
          mockUser
        })
      }))

      const expectedFetchBody = {
        method: 'POST',
        body: JSON.stringify({mockUser}),
        headers: {
          'Content-Type': 'application/json'
        }
      }

      wrapper.instance().submitExistingUser()

      expect(window.fetch).toHaveBeenCalledWith('https://adoptr-be.herokuapp.com/api/v1/sessions', expectedFetchBody)

    });
  });


});