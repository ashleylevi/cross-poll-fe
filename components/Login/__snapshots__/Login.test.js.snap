// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login renders correctly 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "#E5E5E5",
      "flex": 1,
      "fontSize": 50,
      "justifyContent": "space-evenly",
      "paddingBottom": 140,
      "paddingTop": 160,
      "width": "100%",
    }
  }
>
  <View
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "#048BA8",
        "borderRadius": 24,
        "color": "white",
        "flexDirection": "row",
        "fontFamily": "Kohinoor Bangla",
        "fontSize": 50,
        "justifyContent": "center",
        "padding": 10,
        "position": "absolute",
        "shadowColor": "black",
        "shadowOffset": Object {
          "height": 1,
          "width": 0,
        },
        "shadowOpacity": 0.8,
        "shadowRadius": 2,
        "textAlign": "center",
        "top": 60,
        "width": 280,
        "zIndex": 3,
      }
    }
  >
    <Text
      style={
        Object {
          "color": "white",
          "fontFamily": "Kohinoor Bangla",
          "fontSize": 52,
          "textAlign": "center",
        }
      }
    >
      AdoptR
    </Text>
    <View>
      <View
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "transparent",
            "justifyContent": "center",
          }
        }
        theme={
          Object {
            "colors": Object {
              "disabled": "hsl(208, 8%, 90%)",
              "divider": "#bcbbc1",
              "error": "#ff190c",
              "grey0": "#393e42",
              "grey1": "#43484d",
              "grey2": "#5e6977",
              "grey3": "#86939e",
              "grey4": "#bdc6cf",
              "grey5": "#e1e8ee",
              "greyOutline": "#bbb",
              "platform": Object {
                "android": Object {
                  "error": "#f44336",
                  "primary": "#2196f3",
                  "secondary": "#9C27B0",
                  "success": "#4caf50",
                  "warning": "#ffeb3b",
                },
                "ios": Object {
                  "error": "#ff3b30",
                  "primary": "#007aff",
                  "secondary": "#5856d6",
                  "success": "#4cd964",
                  "warning": "#ffcc00",
                },
              },
              "primary": "#2089dc",
              "searchBg": "#303337",
              "secondary": "#8F0CE8",
              "success": "#52c41a",
              "warning": "#faad14",
            },
          }
        }
        underlayColor="white"
      >
        <Text />
      </View>
    </View>
  </View>
  <View
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "white",
        "borderRadius": 16,
        "flex": 1,
        "fontSize": 60,
        "justifyContent": "space-around",
        "marginTop": 20,
        "paddingBottom": 100,
        "paddingTop": 60,
        "shadowColor": "black",
        "shadowOffset": Object {
          "height": 2,
          "width": 1,
        },
        "shadowOpacity": 0.8,
        "shadowRadius": 4,
        "width": "96%",
      }
    }
  >
    <View
      style={
        Object {
          "alignItems": "center",
          "display": "flex",
          "flexDirection": "row",
          "justifyContent": "space-between",
          "position": "absolute",
          "top": 10,
          "width": 260,
          "zIndex": 2,
        }
      }
    >
      <View
        accessible={true}
        isTVSelectable={true}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "opacity": 1,
          }
        }
      >
        <Text
          style={
            Object {
              "color": "grey",
              "fontFamily": "Kohinoor Bangla",
              "fontSize": 32,
            }
          }
        >
          sign in
        </Text>
      </View>
      <View
        accessible={true}
        isTVSelectable={true}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "opacity": 1,
          }
        }
      >
        <Text
          style={
            Object {
              "color": "#048BA8",
              "fontFamily": "Kohinoor Bangla",
              "fontSize": 32,
              "fontWeight": "bold",
              "textDecorationColor": "#048BA8",
              "textDecorationLine": "underline",
            }
          }
        >
          sign up
        </Text>
      </View>
    </View>
    <TextInput
      allowFontScaling={true}
      autoCapitalize="none"
      onChangeText={[Function]}
      placeholder="first name"
      style={
        Object {
          "borderBottomColor": "#048BA8",
          "borderBottomWidth": 2,
          "borderStyle": "solid",
          "fontFamily": "Kohinoor Bangla",
          "fontSize": 20,
          "height": 66,
          "textAlign": "left",
          "width": 320,
        }
      }
      underlineColorAndroid="transparent"
      value=""
    />
    <TextInput
      allowFontScaling={true}
      autoCapitalize="none"
      onChangeText={[Function]}
      placeholder="last name"
      style={
        Object {
          "borderBottomColor": "#048BA8",
          "borderBottomWidth": 2,
          "borderStyle": "solid",
          "fontFamily": "Kohinoor Bangla",
          "fontSize": 20,
          "height": 66,
          "textAlign": "left",
          "width": 320,
        }
      }
      underlineColorAndroid="transparent"
      value=""
    />
    <TextInput
      allowFontScaling={true}
      autoCapitalize="none"
      onChangeText={[Function]}
      placeholder="email"
      style={
        Object {
          "borderBottomColor": "#048BA8",
          "borderBottomWidth": 2,
          "borderStyle": "solid",
          "fontFamily": "Kohinoor Bangla",
          "fontSize": 20,
          "height": 66,
          "textAlign": "left",
          "width": 320,
        }
      }
      type="email"
      underlineColorAndroid="transparent"
      value=""
    />
    <TextInput
      allowFontScaling={true}
      autoCapitalize="none"
      onChangeText={[Function]}
      placeholder="password"
      secureTextEntry={true}
      style={
        Object {
          "borderBottomColor": "#048BA8",
          "borderBottomWidth": 2,
          "borderStyle": "solid",
          "fontFamily": "Kohinoor Bangla",
          "fontSize": 20,
          "height": 66,
          "textAlign": "left",
          "width": 320,
        }
      }
      underlineColorAndroid="transparent"
      value=""
    />
    <TextInput
      allowFontScaling={true}
      onChangeText={[Function]}
      placeholder="confirm password"
      secureTextEntry={true}
      style={
        Object {
          "borderBottomColor": "#048BA8",
          "borderBottomWidth": 2,
          "borderStyle": "solid",
          "fontFamily": "Kohinoor Bangla",
          "fontSize": 20,
          "height": 66,
          "textAlign": "left",
          "width": 320,
        }
      }
      underlineColorAndroid="transparent"
      value=""
    />
    <Text
      style={
        Object {
          "display": "none",
        }
      }
    >
      
    </Text>
  </View>
  <View
    accessible={true}
    isTVSelectable={true}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "backgroundColor": "#048BA8",
        "borderColor": "black",
        "borderRadius": 16,
        "borderStyle": "solid",
        "bottom": 50,
        "opacity": 1,
        "position": "absolute",
        "shadowColor": "black",
        "shadowOffset": Object {
          "height": 2,
          "width": 1,
        },
        "shadowOpacity": 0.8,
        "shadowRadius": 4,
        "width": 280,
      }
    }
  >
    <Text
      style={
        Object {
          "color": "white",
          "fontFamily": "Kohinoor Bangla",
          "fontSize": 32,
          "textAlign": "center",
        }
      }
    >
      submit
    </Text>
  </View>
</View>
`;

exports[`Login should match the snapshot with all data passed in correctly 1`] = `ShallowWrapper {}`;
